Make drivers more consistent / reduce driver API

enumag opened the pull request.

enumag commented on src/BlockingDriver.php:
```diff
@@ -348,7 +348,7 @@ public function chown(string $path, int $uid, int $gid): Promise
         }
 
-        return new Success;
+        return new Success(true);
```
true is returned in other functions like symlink, link and return. Also it seems that other drivers can return a boolean for chown so I added this for consistency.

kelunik replied:
Not sure whether we should change that in a minor. What do other drivers currently return? Could you add a test for this, so all drivers behave equally?

enumag replied:
EioDriver [returns true](https://github.com/amphp/file/blob/9e8bb4defe791f41e56b98193a4399d7083c5ee8/src/EioDriver.php#L347) and UvDriver [returns boolean](https://github.com/amphp/file/blob/master/src/UvDriver.php#L556). Not sure if UvDriver can actually return false or when it would happen.

> Could you add a test for this, so all drivers behave equally?

I'm not exactly sure how. To test chown you first need some user to chown to... also it needs root privileges, doesn't it? I'm guessing that's why there is no test for chown at the moment.

kelunik replied:
I guess we could chown from the current user to the current user?

enumag replied:
Not sure but I guess I can try to add such test to see how that behaves.

enumag replied:
@kelunik Sorry, I kinda forgot about this PR. Added the test just now.

enumag commented on test/DriverTest.php:
```diff
@@ -349,4 +349,13 @@ private function getPermissionsFromStat(array $stat): string
+        $fixtureDir = Fixture::path();
+
+        $original = "{$fixtureDir}/small.txt";
+        $user = fileowner($original);
```
is there a non-blocking way to get the user id of a file owner?

kelunik replied:
Not yet, I think. But doesn't matter for the test.

enumag commented:
ping @kelunik

kelunik commented on src/Driver.php:
```diff
@@ -168,7 +168,7 @@ public function rmdir(string $path): Promise;
      *
      * @param string $path
-     * @return \Amp\Promise
+     * @return \Amp\Promise<array<int, string>>
```
This should probably be `Promise<list<string>>`?

enumag replied:
Do you know if PHPStan understand list type? I'm unsure whether it's psalm's only feature or not.

enumag replied:
ping @ondrejmirtes

enumag replied:
okay, apparently it does: https://phpstan.org/r/e440dbc3-1f34-491f-9629-f0639d98fe4d

enumag replied:
```suggestion
     * @return \Amp\Promise<list<string>>
```

enumag commented on src/functions.php:
```diff
@@ -282,7 +282,7 @@ function rmdir(string $path): Promise
  *
  * @param string $path
- * @return \Amp\Promise<array>
+ * @return \Amp\Promise<array<int, string>>
```
```suggestion
 * @return \Amp\Promise<list<string>>
```

kelunik commented on test/DriverTest.php:
```diff
@@ -349,4 +349,13 @@ private function getPermissionsFromStat(array $stat): string
+        $original = "{$fixtureDir}/small.txt";
+        $user = fileowner($original);
+        $this->assertTrue(yield File\chown($original, $user));
+    }
```
The new test fails for `uv`, could you have a look?

enumag replied:
Okay, I'll check tomorrow. Could you have a look at [this](https://github.com/amphp/sync/pull/10) please? I'd like to finish my work on the AsyncFileMutex tomorrow too.

enumag replied:
@kelunik I tried but it's passing for me locally. Most likely because I'm using an older ext-uv. But I do not wish to install ext-uv 0.3 on localhost yet.

kelunik commented on test/DriverTest.php:
```diff
@@ -349,4 +349,13 @@ private function getPermissionsFromStat(array $stat): string
+        $fixtureDir = Fixture::path();
+
+        $original = "{$fixtureDir}/small.txt";
+        $user = fileowner($original);
```
```suggestion
        $user = \fileowner($original);
```

enumag commented:
@kelunik Most likely chown needs the same treatment with `if ($this->priorVersion) {` as most of the other methods. This might apply to chmod, touch and possibly other methods that are not covered by tests as well.

kelunik commented:
@enumag The test won't work that way due to umask.

enumag commented:
@kelunik Is there a way to fix it or do I have to remove it?

kelunik commented:
@enumag I guess `$old = umask(0)` + `umask($old)` is needed, as it's used here: https://www.php.net/manual/en/function.umask.php

enumag commented:
@kelunik Actually `umask` wasn't the problem. Missing `clearstatcache` was. Second call to `fileperms` returned wrong result.

kelunik commented on test/DriverTest.php:
```diff
@@ -349,4 +349,24 @@ private function getPermissionsFromStat(array $stat): string
+        $fixtureDir = Fixture::path();
+
+        $original = "{$fixtureDir}/small.txt";
+        $this->assertNotSame('0777', substr(sprintf('%o', fileperms($original)), -4));
+        $this->assertTrue(yield File\chmod($original, 0777));
+        clearstatcache();
+        $this->assertSame('0777', substr(sprintf('%o', fileperms($original)), -4));
```
```suggestion
        $this->assertNotSame('0777', \substr(\sprintf('%o', \fileperms($original)), -4));
        $this->assertTrue(yield File\chmod($original, 0777));
        \clearstatcache();
        $this->assertSame('0777', \substr(\sprintf('%o', \fileperms($original)), -4));
```

enumag commented:
Okay the UvDriver is passing now with both ext-uv v0.2 and ext-uv 0.3.

However I'm really not satisfied with this because of #39.

enumag commented:
I'll try to address that in this PR as well. To make things consistent I'll change `symlink`, `link`, `mkdir`, `rmdir`, `rename`, `unlink`, `chown`, `chmod` and `touch` to `Promise<null>` + throw on failure for all drivers. That seems to be the intended behavior according to the docblocks before this PR even though the drivers behave differently.

`Promise<bool>` makes sense for `exists`, `isfile`, `isdir`.

Whether you feel this needs a new major release is up to you. I consider it a bugfix. It needs to be fixed either way.

enumag commented:
I updated the docblocks and refactored DriverTest accordingly + added tests for operation failures.

Now I need to fix all the drivers.

enumag commented:
@kelunik @trowski Okay, this should be everything. I can also confirm that the tests are passing with ext-uv 0.2 since I'm using it locally. Please review.

enumag commented:
Tried to add a [test for hardlink creation](https://github.com/amphp/file/pull/38/commits/24910de18e73485d43fc9ae90a23177526710d8d) but there seems to be some issue with UvDriver - the test doesn't end. Any idea what's going on here?

kelunik commented:
@enumag Thank you for your work on this! I think it makes sense to release this a v2.0, because driver implementations were inconsistent, but as long as you always used the same driver, it always worked the same. Changing it now would potentially break applications.

Travis CI also builds a job with uv v0.2.2 now, so we'll notice whether it works or fails, but thanks for checking locally.

I guess the link case might need to clear the stat cache?

enumag commented:
> I think it makes sense to release this a v2.0, because driver implementations were inconsistent, but as long as you always used the same driver, it always worked the same. Changing it now would potentially break applications.

Yeah, that's reasonable. Sounds good to me.

> Travis CI also builds a job with uv v0.2.2 now, so we'll notice whether it works or fails, but thanks for checking locally.

Oh... I have not noticed. Good job!

> I guess the link case might need to clear the stat cache?

No. If it was stat cache then it would simply return wrong result but it wouldn't hang the test like this... We're either calling it wrong or there is a bug in ext-uv. I can't tell which is it since there is no documentation nor examples how ext-uv functions should be used so I don't know.

kelunik commented on src/Driver.php:
```diff
@@ -216,7 +216,7 @@ public function get(string $path): Promise;
      * @param string $path The file path to which to $contents should be written
      * @param string $contents The data to write to the specified $path
-     * @return \Amp\Promise A promise resolving to the integer length written upon success
+     * @return Promise<int> A promise resolving to the integer length written upon success
```
I guess this should also return `Promise<void>`

enumag replied:
Why? `A promise resolving to the integer length written upon success` - you would lose that information. Though maybe it's just always the same as `strlen($contents)` and that's why you wanna drop it?

kelunik replied:
Yes, exactly, it's just `strlen` on the input.

kelunik commented on test/DriverTest.php:
```diff
@@ -85,7 +115,7 @@ public function testLstat(): \Generator
         $target = "{$fixtureDir}/small.txt";
         $link = "{$fixtureDir}/symlink.txt";
-        $this->assertTrue(yield File\symlink($target, $link));
+        yield File\symlink($target, $link);
```
This could be changed to `$this->assertNull`

enumag replied:
I'm using assertNull in testSymlink. This is not a test for the symlink method so I omitted it on purpose. Do you want it here anyway?

kelunik replied:
No, fine then.

kelunik commented on test/DriverTest.php:
```diff
@@ -349,4 +453,44 @@ private function getPermissionsFromStat(array $stat): string
+        $path = "{$fixtureDir}/small.txt";
+        $this->assertNotSame('0777', \substr(\sprintf('%o', \fileperms($path)), -4));
+        $this->assertNull(yield File\chmod($path, 0777));
+        \clearstatcache();
```
Shouldn't `chmod` do the `clearstatcache()`?

enumag replied:
It was there for `fileperms()`  to work correctly. But maybe I can use `Amp\File\stat` to get them instead... I'll try.

kelunik commented on src/functions.php:
```diff
@@ -140,6 +140,32 @@ function isfile(string $path): Promise
+ * @fails \Amp\Files\FilesystemException If the path does not exist
+ * @return Promise<string>
+ */
+function fileperms(string $path): Promise
```
I'm not sure whether we should keep the PHP names or should use better names for each function.

enumag replied:
By keeping the same name you make it obvious that this function is an async equivalent. But if you want to name it differently, I'm fine with that.

enumag commented:
@trowski @kelunik I'm considering adding these [functions](https://github.com/amphp/file/pull/38/commits/504866917b55af8f7fe18fe1226b774c6965e094) since it's quite difficult to get this info from `stat`. What do you think?

kelunik commented:
@enumag Seems fine. If it's possible but difficult, should we based these functions on `stat` instead of adding them to `Driver`?

enumag commented:
@kelunik Depends... Some other things like isfile and isdir could also be based on stat but aren't...

enumag commented:
Note for myself: Review which functions change anything that is retrieved using stats (like `chown`, `chmod`, `touch`) and call `StatCache::clear()` in them. Assert that the cache has been cleared in DriverTest.

enumag commented:
@kelunik So what do you think? Should I refactor some of the functions to do just additional logic with the result of `DriverInterface::stat()` and simplify the drivers by removing them there? This would affect `exists`, `isdir`, `isfile`, `size`, `mtime`, `atime` and `ctime` + the new `islink` and `fileperms` (?) functions.

kelunik commented:
I personally think that's a good, as it reduces the code specific for each driver and thus the potential for inconsistencies, but I'd like to hear @trowski's opinion on that as well.

trowski commented:
I agree that's a good idea. That's effectively what EIO and UV already do, so simplifying even further makes a lot of sense to me.

enumag commented:
@kelunik @trowski All done but `UvDriver::link()` is still broken. I have no idea why.

enumag commented:
TODO: add tests for islink

enumag commented:
@kelunik @trowski I fixed isfile, isdir, isSymlink and chown... But still no idea what's wrong with `UvDriver::link()`.

enumag commented on src/functions.php:
```diff
@@ -132,58 +167,128 @@ function isdir(string $path): Promise
+
+    lstat($path)->onResolve(function (?Throwable $error, ?array $result) use ($deferred): void {
+        if ($result !== null) {
+            $deferred->resolve((bool) ($result["mode"] & 0020000));
```
Not entirely sure about this. According to [stat docs](https://www.php.net/manual/en/function.stat.php), link is `0120000`. But using that value here fails the tests.

kelunik replied:
A `(bool)` cast here isn't enough, you need to check `($mask & $mode) === $mask`, see https://3v4l.org/mbCiU. Does that fix the test?

enumag replied:
It does! Thanks, I'm not used to bitmask handling. This is good to know.

kelunik commented on composer.json:
```diff
@@ -40,6 +40,7 @@
     },
     "require-dev": {
         "ext-eio": "^2",
+        "ext-posix": "*",
```
We shouldn't require `ext-posix` here, because we also support Windows.

enumag replied:
Should I remove the fifo files from tests then?

kelunik replied:
There's no reason to remove tests, we just need to skip them if `ext-posix` isn't available.

kelunik commented on src/BlockingDriver.php:
```diff
@@ -322,15 +187,21 @@ public function scandir(string $path): Promise
      * {@inheritdoc}
      */
-    public function chown(string $path, int $uid, int $gid): Promise
+    public function chown(string $path, ?int $uid, ?int $gid = null): Promise
```
Should we have a default for `$gid` at all?

enumag replied:
Yeah, I was considering that too... I kinda think we should keep the default in the chown function but drop it in DriverInterface. Do you agree?

kelunik replied:
Yes, seems fine.

enumag replied:
I applied the same treatment to the null defaults of `Driver::touch()` as well. Not sure if I should do it for `Driver::mkdir()` too...

kelunik commented on src/BlockingDriver.php:
```diff
@@ -348,17 +219,29 @@ public function chown(string $path, int $uid, int $gid): Promise
         }
 
-        return new Success;
+        StatCache::clear();
```
This should probably pass the `$path`?

kelunik commented on src/BlockingDriver.php:
```diff
@@ -348,17 +219,29 @@ public function chown(string $path, int $uid, int $gid): Promise
+            return new Failure(new FilesystemException($message));
+        }
+
+        StatCache::clear();
```
This should probably pass the `$path`?

kelunik commented on src/EioDriver.php:
```diff
@@ -436,6 +306,21 @@ public function mkdir(string $path, int $mode = 0777, bool $recursive = false):
+
+        $this->stat($path)->onResolve(function ($error, $result) use ($deferred): void {
+            if ($result) {
+                $deferred->resolve(!($result["mode"] & \EIO_S_IFREG));
```
```suggestion
                $deferred->resolve(($result["mode"] & \EIO_S_IFDIR) === \EIO_S_IFDIR);
```

enumag replied:
@kelunik The constant `EIO_S_IFDIR` does not exist: https://www.php.net/manual/en/eio.constants.php

kelunik replied:
I guess that's something we should add then? https://github.com/rosmanov/pecl-eio/blob/57e1f529c2950ed41d46b559c72a37ca1a46ee97/php7/php_eio.c#L1071

enumag replied:
I don't think so. eio_mknod (which these constants are meant for) creates a normal file (https://www.php.net/manual/en/function.eio-mknod.php), for directory there is a https://www.php.net/manual/en/function.eio-mkdir.php

kelunik commented on src/UvDriver.php:
```diff
@@ -471,14 +320,27 @@ public function mkdir(string $path, int $mode = 0777, bool $recursive = false):
+
+        $this->stat($path)->onResolve(function ($error, $result) use ($deferred): void {
+            if ($result) {
+                $deferred->resolve(!($result["mode"] & \UV::S_IFREG));
```
```suggestion
                $deferred->resolve(($result["mode"] & \UV::S_IFDIR) === \UV::S_IFDIR);
```

enumag commented:
@bwoebi Could you take a look what's wrong with `UvDriver::link()` please? It still hangs the tests and I can't figure it out. I actually kinda suspect there might be something wrong in ext-uv.

kelunik commented on src/UvDriver.php:
```diff
@@ -358,13 +219,7 @@ public function link(string $target, string $link): Promise
-
-            $deferred->resolve((bool) $fh);
-        });
+        \uv_fs_link($this->loop, $target, $link, \UV::S_IRWXU | \UV::S_IRUSR, $this->createGenericCallback($deferred, "Could not create hard link"));
```
```suggestion
        \uv_fs_link($this->loop, $target, $link, $this->createGenericCallback($deferred, "Could not create hard link"));
```

kelunik commented on test/Fixture.php:
```diff
@@ -29,11 +29,38 @@ public static function init(): void
+                "Failed creating temporary test fixture symlink loop"
+            );
+        }
+        if (extension_loaded('posix')) {
```
```suggestion
        if (\extension_loaded('posix')) {
```

kelunik commented on test/DriverTest.php:
```diff
@@ -156,46 +187,76 @@ public function testSizeFailsOnDirectoryPath(): \Generator
+        yield 'filelink' => [false, 'filelink'];
+        yield 'dir' => [true, 'dir'];
+        yield 'dirlink' => [true, 'dirlink'];
+        if (extension_loaded('posix')) {
```
```suggestion
        if (\extension_loaded('posix')) {
```

kelunik commented on test/DriverTest.php:
```diff
@@ -156,46 +187,76 @@ public function testSizeFailsOnDirectoryPath(): \Generator
+        yield 'filelink' => [true, 'filelink'];
+        yield 'dir' => [false, 'dir'];
+        yield 'dirlink' => [false, 'dirlink'];
+        if (extension_loaded('posix')) {
```
```suggestion
        if (\extension_loaded('posix')) {
```

kelunik commented on test/DriverTest.php:
```diff
@@ -156,46 +187,76 @@ public function testSizeFailsOnDirectoryPath(): \Generator
+        yield 'filelink' => [true, 'filelink'];
+        yield 'dir' => [false, 'dir'];
+        yield 'dirlink' => [true, 'dirlink'];
+        if (extension_loaded('posix')) {
```
```suggestion
        if (\extension_loaded('posix')) {
```

kelunik commented:
@enumag I've fixed `UvDriver::link`, it had a parameter too much.

@trowski Please review, I'd like to finally merge this.

enumag commented:
@kelunik @trowski Could we have a release with these changes?

kelunik commented:
@enumag Yes, we should indeed finally tag v2. I guess the code is ready, we just need to write a changelog. I originally wanted to wrap `stat` into an object, but probably won't do that. Do you want to help with the changelog?

enumag commented:
@kelunik Well the notable changes are these:

```
- Added AsyncFileMutex (#43)
- Consistent and reduced drivers API (#38)
- Redesigned drivers API (#53)
```

Then it depends on if you want to
- list all the renamed functions from your API redesign
- list which methods were removed from drivers when reducing their API
- describing the philosophy behind consistent API (https://github.com/amphp/file/pull/38#issuecomment-622939210)

kelunik commented:
That's a good starting point, yes. I think listing the renames makes senses, as it only needs to be done once, otherwise everybody will have to spend time looking things up. I'm not sure whether detailed changelogs for the driver interface makes sense, as there probably aren't many people who implemented their own driver. Do you want to work on that?

enumag commented:
@kelunik here is what I gathered:

Renamed:
open -> openFile
stat -> getStatus
lstat -> getLinkStatus
size -> getSize
isdir -> isDirectory
isfile -> isFile
mtime -> getModificationTime
atime -> getAccessTime
ctime -> getCreationTime
symlink -> createSymlink
link -> createHardlink
readlink -> resolveSymlink
rename -> move
unlink -> deleteFile
rmdir -> deleteDirectory
scandir -> listFiles
chmod -> changePermissions
get -> read
put -> write

Added:
isSymlink

Changed:
mkdir split to createDirectory and createDirectoryRecursively
chown renamed to changeOwner, $uid and $gid now accept null instead of -1 to ignore

enumag commented:
ping @kelunik Anything else I can help with here?

kelunik commented:
No, thank you! I'll get it tagged in the coming days. Sorry about the delay.

kelunik commented:
I finally got to my notebook and tagged the release, thank you so much!

enumag commented:
Thanks @kelunik!

I noticed a few things when browsing the latest commits.

1) You're apparently using a dev version of ext-uv (https://github.com/amphp/file/commit/f45eb1177f910924151bacf4aea28161ef2fcaa9). How are things looking there? Could we have a new release?

2) Also noticed that you're ignoring ext-eio on PHP 8 (https://github.com/amphp/file/commit/90f955076ce24e7bb629029be71ec527c7afb143). However iirc there is already a PHP 8 compatible RC release of ext-eio.

kelunik commented:
I'm using the current master branch of ext-uv, yes. I'm not sure what changes from the last release there.

I'm ignoring eio, because the setup-php script doesn't install it currently. If you have a fix for that, please send a PR.
